autodiff/src/LinSigmoid.cpp

```diff
@@ -1,58 +1,70 @@
-/*
- * LinSigmoid.cpp
- *
- *  Created on: Jul 17, 2014
- *      Author: psp
- */
-
 #include "LinSigmoid.h"
 
-#include "TermBuilder.h"
 #include "Constant.h"
+#include "TermHolder.h"
 
 #include <cmath>
 #include <limits>
+#include <sstream>
+
+namespace autodiff
+{
+LinSigmoid::LinSigmoid(TermPtr arg, TermHolder* owner)
+    : UnaryFunction(arg, owner)
+{
+}
 
-namespace autodiff {
-LinSigmoid::LinSigmoid(shared_ptr<Term> arg)
-        : Term() {
-    this->arg = arg;
+int LinSigmoid::accept(ITermVisitor* visitor)
+{
+    return visitor->visit(this);
 }
 
-int LinSigmoid::accept(shared_ptr<ITermVisitor> visitor) {
-    shared_ptr<LinSigmoid> thisCasted = dynamic_pointer_cast<LinSigmoid>(shared_from_this());
-    return visitor->visit(thisCasted);
+void LinSigmoid::acceptRecursive(ITermVisitor* visitor)
+{
+    _arg->acceptRecursive(visitor);
+    visitor->visit(this);
 }
 
-shared_ptr<Term> LinSigmoid::aggregateConstants() {
-    arg = arg->aggregateConstants();
-    if (dynamic_pointer_cast<Constant>(arg) != 0) {
-        shared_ptr<Constant> arg = dynamic_pointer_cast<Constant>(arg);
-        double e = exp(-arg->value);
-        if (e == numeric_limits<double>::infinity()) {
-            return TermBuilder::constant(Term::EPSILON);
+TermPtr LinSigmoid::aggregateConstants()
+{
+    _arg = _arg->aggregateConstants();
+    if (_arg->isConstant()) {
+        double e = exp(-static_cast<Constant*>(_arg)->getValue());
+        if (e == std::numeric_limits<double>::infinity()) {
+            return _owner->constant(Term::EPSILON);
         } else {
             e = 1.0 / (1.0 + e);
         }
         if (e < Term::EPSILON) {
-            return TermBuilder::constant(Term::EPSILON);
+            return _owner->constant(Term::EPSILON);
         } else {
-            return TermBuilder::constant(e);
+            return _owner->constant(e);
         }
-    } else {
-        return shared_from_this();
     }
+    return this;
 }
 
-shared_ptr<Term> LinSigmoid::derivative(shared_ptr<Variable> v) {
-    return arg->derivative(v);
+TermPtr LinSigmoid::derivative(VarPtr v) const
+{
+    // TermPtr e = _owner->exp(-arg);
+    // return e / _owner->constantPower(e + 1.0, 2.0);
+    return _arg->derivative(v);
 }
 
-string LinSigmoid::toString() {
-    string str;
-    str.append("sigmoid( ");
-    str.append(arg->toString());
-    str.append(" )");
-    return str;
+std::string LinSigmoid::toString() const
+{
+    std::stringstream str;
+    str << "sigmoid( " << _arg->toString() << " )";
+    return str.str();
+}
+
+void LinSigmoid::Eval(const Tape& tape, const Parameter* params, double* result, const double* vars, int dim)
+{
+    const double* arg = tape.getValues(params[0].asIdx);
+    const double e = exp(-arg[0]);
+    result[0] = 1.0 / (1.0 + e);
+    for (int i = 1; i <= dim; ++i) {
+        result[i] = arg[i];
```

Comment:
purposefully. The derivate of the linsigmoid is 1*inner.
The Sigmoid implements the more normal one. This done to avoid numerical issue in the presence of many finely tuned sigmoids.

Reply:
Adding a comment